### ISSUE TITLE
How to originate calls from service

### ISSUE THREAD

**Issue by enp** [friends-of-freeswitch/switchio#61]
Hi,
Is it possible to originate call from my application (I mean my class with @app and @event_callback)?
Maybe anybody can show me simple service based on https://switchio.readthedocs.io/en/latest/services.html which can originate calls in addition to receiving incomimg calls?

**Comment by tgoodlet** [comment]
Hey @enp.

Hmm I haven't really done this before (it's usually the opposite - originate a call and then use an app to process it). One thing to start with is the [quickstart](https://switchio.readthedocs.io/en/latest/quickstart.html) which shows how to manually originate new calls.

It's actually a bit of an odd use case since apps are `Session` oriented while an [originate cmd](https://freeswitch.org/confluence/display/FREESWITCH/mod_commands#mod_commands-originate) spawns a new (set of) session(s), but I guess you could try a few things:
- use the `Session.con` which is a ref to the [`Connection` api](https://github.com/friends-of-freeswitch/switchio/blob/master/switchio/connection.py#L111) which has methods for executing ESL api calls manually (you'll notice the [`Session`](https://github.com/friends-of-freeswitch/switchio/blob/master/switchio/models.py#L82) and [`Client`](https://github.com/friends-of-freeswitch/switchio/blob/master/switchio/api.py#L23) delegate to this internally). 
- use the `Service.pool.clients` list to access a [`Client`](https://github.com/friends-of-freeswitch/switchio/blob/master/switchio/api.py#L23) which defines the high level [originate method](https://github.com/friends-of-freeswitch/switchio/blob/master/switchio/api.py#L351) (also there's a small tutorial on [how to originate with a `Client`](https://switchio.readthedocs.io/en/latest/usage.html#example-snippet)).

Note this is all a bit tricky with a `Service` since it's meant to manage a FS process cluster; knowing *which* FS process should originate the call is potentially involved to figure out. This is exactly why `switchio` has it's own built in [autodialer](https://github.com/friends-of-freeswitch/switchio/blob/master/switchio/apps/call_gen.py#L113) for this kind of thing.

Another thing to note is that the `@event_callback` api is the legacy callback system. You can now use `@switchio.coroutine` to decorate async functions. Unfortunately this hasn't been documented yet (as well as many other nice things) as per #33...

Anyway hope this helps.

**Comment by enp** [comment]
Thanks for you response, but I see @event_callback (and maybe @app) is outdated, right? 

Anyway I need for a bit different. I need for some service with two entry point:

- SIP - it must answer/playback/hangup on incoming call
- HTTP - it must originate/playback/hangup on incoming HTTP request

So, I need for some mix of https://aiohttp.readthedocs.io/en/stable/web_quickstart.html#run-a-simple-web-server and https://github.com/friends-of-freeswitch/switchio#use-the-power-of-async-and-await with one event loop for two entry points instead of web.run_app(app) and service.run()

What is the best way to do it?

**Comment by enp** [comment]
Suppose I already have [simple web service](https://gist.github.com/enp/abad4cbc3e5c6c9821708845c53d6add) with explicit event loop. I need to define switchio EventLoop and Connection both in __init__, right? How to define FS event handlers in this case?

**Comment by tgoodlet** [comment]
To start the `Service` running in a background thread you could simply [start the underlying clients](https://github.com/friends-of-freeswitch/switchio/blob/master/switchio/serve.py#L29) but without waiting on them as is done in `Service.run()`.

**Comment by tgoodlet** [comment]
Hmm I guess it would be interesting to see what happens if you tried to use one loop for both.
I know we can pass in a `loop` to the `EventLoop` but I don't think we have any tests really for that.

**Comment by enp** [comment]
So, it is impossible to use one event loop for both now?

Btw, I tried just to import EventListener or EventLoop but I got error:

```
Traceback (most recent call last):
  File "examples/freeswitch/call.py", line 6, in <module>
    from switchio import Client, EventListener, EventLoop
ImportError: cannot import name 'EventListener'
```
Why I can't do this?

**Comment by tgoodlet** [comment]
>So, it is impossible to use one event loop for both now?

No, it just might not work reliably and it's never been tested.

>Why I can't do this?

Because `EventListener` isn't exposed at the top level package.
I'd recommend reading the source code before just blindly importing names.

If you are having problems unrelated to this original issue please instead join our [riot chat]( https://riot.im/app/#/room/#freeswitch:matrix.org).

**Comment by tgoodlet** [comment]
@enp ahh I just realized,

>Why I can't do this?

Yeah so [the internals tutorial](https://switchio.readthedocs.io/en/latest/usage.html#example-snippet) is obviously out of date. 

Like I said docs are in a shambles since the port to `asyncio`.. see #33 as well.

I made an issue #62.

**Comment by tgoodlet** [comment]
@enp oh nice. 
Glad to see the low level stuff is at least useful :+1: 

If you're interested in exposing what you need in some of the higher level APIs we'll gladly take PRs for it!

I know that getting this all working with a single loop / thread is definitely possible I just haven't played with it. I've created #66 to track.

**Comment by enp** [comment]
Sure, it will be interesting for me to see higher level example with switchio-specific decorators for event processing and aiohttp-specific decorators for http requests/responses both instead of working with loops manually :)
